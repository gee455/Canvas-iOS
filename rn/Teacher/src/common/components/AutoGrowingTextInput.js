//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { PureComponent } from 'react'
import { TextInput } from 'react-native'

type State = { height: number }
type Props = {
  defaultHeight: number,
  style?: any,
  onContentSizeChange?: Function,
}

export default class AutoGrowingTextInput extends PureComponent<Props, State> {
  state: State = { height: this.props.defaultHeight }

  updateContentSize = (e: any) => {
    // By adding a few pixels to the height of this content size, it ensures that the view will not scroll
    this.setState({ height: Math.max(e.nativeEvent.contentSize.height + 5, this.props.defaultHeight) })
    this.props.onContentSizeChange && this.props.onContentSizeChange(e)
  }

  render () {
    return (
      <TextInput
        {...this.props}
        style={[this.props.style, { height: this.state.height }]}
        multiline
        onContentSizeChange={this.updateContentSize}
      />
    )
  }
}
