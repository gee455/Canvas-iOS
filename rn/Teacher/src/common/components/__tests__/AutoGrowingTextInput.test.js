//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// @flow

import React from 'react'
import AutoGrowingTextInput from '../AutoGrowingTextInput'
import renderer from 'react-test-renderer'

let defaultProps = {
  defaultHeight: 54,
  onContentSizeChange: jest.fn(),
}

describe('AutoGrowingTextInput', () => {
  beforeEach(() => jest.resetAllMocks())
  it('renders', () => {
    let tree = renderer.create(
      <AutoGrowingTextInput {...defaultProps} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('adjusts the height when content size is updated', () => {
    let tree = renderer.create(
      <AutoGrowingTextInput {...defaultProps} />
    )

    let event = { nativeEvent: { contentSize: { height: 100 } } }

    tree.getInstance().updateContentSize(event)
    expect(tree.toJSON()).toMatchSnapshot()
    expect(defaultProps.onContentSizeChange).toHaveBeenCalledWith(event)
  })
})
