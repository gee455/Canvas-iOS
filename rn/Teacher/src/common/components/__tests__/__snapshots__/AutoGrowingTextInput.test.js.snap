// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoGrowingTextInput adjusts the height when content size is updated 1`] = `
<TextInput
  allowFontScaling={true}
  defaultHeight={54}
  multiline={true}
  onContentSizeChange={[Function]}
  style={
    Array [
      undefined,
      Object {
        "height": 105,
      },
    ]
  }
/>
`;

exports[`AutoGrowingTextInput renders 1`] = `
<TextInput
  allowFontScaling={true}
  defaultHeight={54}
  multiline={true}
  onContentSizeChange={[Function]}
  style={
    Array [
      undefined,
      Object {
        "height": 54,
      },
    ]
  }
/>
`;
